### PR TITLE
Add filter for exhaustive switch expression

### LINKED ISSUE #771: Add filter for exhaustive switch-expression
Currently for exhaustive switch-expression ([JEP 361](https://openjdk.java.net/jeps/361))

```java
public class SwitchExpressionTarget {

	private enum E {
		TRUE, FALSE,
	}

	private static int target(E e) {
		return switch (e) {
			case TRUE -> 1;
			case FALSE -> 0;
		};
	}

	public static void main(String[] args) {
		target(E.TRUE);
		target(E.FALSE);
	}

}
```

with JDK 14 EA b29

```
$ java --version
openjdk version "14-ea" 2020-03-17
OpenJDK Runtime Environment (build 14-ea+29-1384)
OpenJDK 64-Bit Server VM (build 14-ea+29-1384, mixed mode, sharing)

$ javac --release 14 SwitchExpressionTarget.java

$ java -javaagent:jacoco-0.8.5/lib/jacocoagent.jar SwitchExpressionTarget 

$ java -jar jacoco-0.8.5/lib/jacococli.jar \
    report jacoco.exec \
    --classfiles SwitchExpressionTarget.class \
    --sourcefiles . \
    --html report
```

we produce

report

branch that throws `IncompatibleClassChangeError` should be filtered out

```
  private static int target(SwitchExpressionTarget$E);
    descriptor: (LSwitchExpressionTarget$E;)I
    flags: (0x000a) ACC_PRIVATE, ACC_STATIC
    Code:
      stack=2, locals=2, args_size=1
         0: getstatic     #2                  // Field SwitchExpressionTarget$1.$SwitchMap$SwitchExpressionTarget$E:[I
         3: aload_0
         4: invokevirtual #3                  // Method SwitchExpressionTarget$E.ordinal:()I
         7: iaload
         8: lookupswitch  { // 2
                       1: 36
                       2: 41
                 default: 46
            }
        36: iconst_1
        37: istore_1
        38: goto          54
        41: iconst_0
        42: istore_1
        43: goto          54
        46: new           #4                  // class java/lang/IncompatibleClassChangeError
        49: dup
        50: invokespecial #5                  // Method java/lang/IncompatibleClassChangeError."<init>":()V
        53: athrow
        54: iload_1
        55: ireturn
      LineNumberTable:
        line 8: 0
        line 9: 36
        line 10: 41
        line 8: 55
```

### PR DESCRIPTION
**Pull request by Godin**
Before this change for the following `Example.java`

```
enum Example {
    A, B;

    static int example(Example e) {
        return switch (e) {
        case A -> 1;
        case B -> 2;
        };
    }

    public static void main(String[] args) {
        example(A);
        example(B);
    }
}
```

using

```
openjdk version "17.0.2" 2022-01-18 LTS
OpenJDK Runtime Environment Zulu17.32+13-CA (build 17.0.2+8-LTS)
OpenJDK 64-Bit Server VM Zulu17.32+13-CA (build 17.0.2+8-LTS, mixed mode, sharing)
```

execution of

```
javac Example.java -d classes
java -javaagent:jacocoagent.jar -cp classes Example
java -jar jacococli.jar report jacoco.exec --classfiles classes --sourcefiles . --html report
```

produces

before

before

and after this change

after

after

---

Fixes #771

### CODE REVIEW

File: org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/ExhaustiveSwitchFilter.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+				dflt = ((TableSwitchInsnNode) start).dflt;
+				labels = ((TableSwitchInsnNode) start).labels;
+			} else {
+				return;
+			}
+
+			cursor = skipToLineNumberOrInstruction(dflt);
+			if (cursor == null) {
+				return;
+			} else if (cursor.getType() == AbstractInsnNode.LINE) {
```
Comment (marchof) [inline org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/ExhaustiveSwitchFilter.java:63]: In consistency with other early returns I would use else here.
```suggestion
			}
			if (cursor.getType() == AbstractInsnNode.LINE) {
```

File: org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/ExhaustiveSwitchFilter.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+				if (line != ((LineNumberNode) cursor).line) {
+					return;
+				}
+				cursor = skipNonOpcodes(cursor);
+			}
+			if (cursor == null || cursor.getOpcode() != Opcodes.NEW) {
+				return;
+			}
+			if ("java/lang/MatchException"
```
Comment (marchof) [inline org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/ExhaustiveSwitchFilter.java:72]: What about some inline documentation or extraction into two method with speaking names.

    String type = ((TypeInsnNode) cursor).desc;
    if (!isJava5Exception(desc) && !isJava21Exception(desc) {
      return;
    }

File: org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/ExhaustiveSwitchFilter.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+						"()V");
+			} else {
+				return;
+			}
+			nextIs(Opcodes.ATHROW);
+			if (cursor == null) {
+				return;
+			}
+			output.ignore(dflt, cursor);
+			final HashSet<AbstractInsnNode> replacements = new HashSet<AbstractInsnNode>();
```
Comment (marchof) [inline org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/ExhaustiveSwitchFilter.java:93]: Use more general type `Set` here?

**Review by marchof — APPROVED** [review]
Great work again! I left some formal comments you might want to consider.

File: org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/ExhaustiveSwitchFilter.java (hunk 1 above)
Comment (marchof) [reply org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/ExhaustiveSwitchFilter.java]: That's what I wanted to say. The truth is in the code 😅

File: org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/ExhaustiveSwitchFilter.java (hunk 3 above)
Comment (marchof) [reply org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/ExhaustiveSwitchFilter.java:96]: Ok, one day in future we will replace it with `var` anyways 😄

**Comment by nioncode** [comment]
@Godin should this already handle multi-line switch expressions? I still get missed branch warnings in my code, where I have something like:

```java

enum Example {
    A, B;

    static int example(Example e) {
        return switch (e) {
        case A -> {
            if (false) {
              throw new InvalidStateException();
            }
            yield 1
        };
        case B -> {
            if (false) {
              throw new InvalidStateException();
            }
            yield 2;
        };
    }

    public static void main(String[] args) {
        example(A);
        example(B);
    }
}
```

**Comment by Godin** [comment]
@nioncode for the following `Example.java` which is the same example as yours after fixing compilation errors

```
enum Example {
    A, B;

    static int example(Example e) {
        return switch (e) {
        case A -> {
            if (false) {
              throw new IllegalStateException();
            }
            yield 1;
        }
        case B -> {
            if (false) {
              throw new IllegalStateException();
            }
            yield 2;
        }
        };
    }

    public static void main(String[] args) {
        example(A);
        example(B);
    }
}
```

using

```
openjdk version "21" 2023-09-19
OpenJDK Runtime Environment (build 21+35-2513)
OpenJDK 64-Bit Server VM (build 21+35-2513, mixed mode, sharing)
```

execution of

```
javac Example.java -d classes
java -javaagent:jacoco-0.8.12/lib/jacocoagent.jar -cp classes Example
java -jar jacoco-0.8.12/lib/jacococli.jar report jacoco.exec --classfiles classes --sourcefiles . --html report
```

produces the following report

Screenshot 2024-04-29 at 11 38 05

Screenshot 2024-04-29 at 11 37 46

**Comment by nioncode** [comment]
This is true. Sorry, I ran this through gradle and it seems it does not pick up the latest version by default.

Just for others, if you use the jacoco plugin in gradle, you can force a specific version via:
```
plugins {
  id "jacoco" // note: don't specify a version here, specify it in the jacoco block below
}

jacoco {
  toolVersion = "0.8.12"
}
```

Everything works fine as is, sorry for the noise.

**Pull request merged**